sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -315,7 +322,80 @@ public override long Position
 
         public override long Seek(long offset, SeekOrigin origin)
         {
-            throw new NotSupportedException();
+            long newPosition = CalculateNewPosition(offset, origin);
+
+            if (newPosition == _position)
+            {
+                return _position;
+            }
+
+            // newPosition < 0
+            if (newPosition < 0)
+            {
+                throw new ArgumentException($"New {nameof(offset)} cannot be less than 0.  Value was {newPosition}");
+            }
+
+            // newPosition > _length
+            if (newPosition > _length)
+            {
+                if (_allowBlobModifications)
+                {
+                    _position = newPosition;
+                    _seeked = true;
+                    return newPosition;
+                }
+                else
+                {
+                    throw new ArgumentException(
+                        $"{nameof(newPosition)} exceeds known blob or file length.  This condition is not allowed with allowBlobModifications == false.");
+                }
+            }
+
+            // newPosition is less than _position, but within _buffer.
+            long beginningOfBuffer = _position - _bufferPosition;
+            if (newPosition < _position && newPosition > beginningOfBuffer)
+            {
+
+                _bufferPosition = (int)(newPosition - beginningOfBuffer);
+                _position = newPosition;
+                return newPosition;
+            }
+
+            // newPosition is greater than _position, but within _buffer.
+            long endOfBuffer = _position + (_bufferLength - _bufferPosition);
+            if (newPosition > _position && newPosition < endOfBuffer)
+            {
+                _bufferPosition = (int)(newPosition - beginningOfBuffer);
+                _position = newPosition;
+                return newPosition;
+            }
+
+            // newPosition is outside of _buffer, we will need to re-download.
+            _seeked = true;
+            _position = newPosition;
+            return newPosition;
+        }
+
+        internal long CalculateNewPosition(long offset, SeekOrigin origin)
+        {
+            switch (origin)
+            {
+                case SeekOrigin.Begin:
+                    return offset;
+                case SeekOrigin.Current:
+                    return _position + offset;
+                case SeekOrigin.End:
+                    if (_allowBlobModifications)
+                    {
+                        throw new ArgumentException($"Cannot {nameof(Seek)} with {nameof(SeekOrigin)}.{nameof(SeekOrigin.End)} on a growing blob or file.");
```

Comment:
or we could use current length to calculate this (extra get properties call).

Reply:
I'm not sure if we can, Seek() is a sync operation, and there doesn't appear to be a SeekAsync() - https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.seek?view=netcore-3.1.

Based on this assumption, I only updated the positions in the Seek() call, and do the necessary network calls in ReadInternal().

Reply:
makes sense. maybe we should disallow all seeks if blob is modifiable.